extensions/positron-python/src/client/positron/session.ts

```diff
@@ -527,66 +518,63 @@ export class PythonRuntimeSession implements positron.LanguageRuntimeSession, vs
-                if (this._kernel) {
-                    this._kernel.emitJupyterLog(`Stopping Positron LSP server`);
-                }
-                await this._lsp?.deactivate(awaitStop);
-            });
-        } else {
-            return undefined;
-        }
-    }
+            if (this._lsp.state !== LspState.stopped && this._lsp.state !== LspState.uninitialized) {
+                // LSP is already started.
+                return;
+            }
+
+            if (!this._kernel) {
+                traceWarn('Cannot start LSP; kernel not started');
+                return;
+            }
 
-    private async startLsp(): Promise<void> {
-        if (this._kernel) {
             this._kernel.emitJupyterLog('Starting Positron LSP server');
 
             // Create the LSP comm, which also starts the LSP server.
             // We await the server selected port (the server selects the
             // port since it is in charge of binding to it, which avoids
             // race conditions). We also use this promise to avoid restarting
             // in the middle of initialization.
-            this._lspStarting = this._kernel.startPositronLsp('127.0.0.1');
-            const port = await this._lspStarting;
+            const port = await this._kernel.startPositronLsp('127.0.0.1');
 
             this._kernel.emitJupyterLog(`Starting Positron LSP client on port ${port}`);
 
-            await this._lsp?.activate(port);
-        }
+            await this._lsp.activate(port);
+        });
+    }
+
+    /**
+     * Stops the LSP if it is running
+     *
+     * @returns A promise that resolves when the LSP has been deactivated.
+     */
+    async deactivateLsp(): Promise<void> {
+        return this._lspQueue.add(async () => {
+            if (!this._lsp || this._lsp.state !== LspState.running) {
+                // LSP is already stopped.
+                return;
+            }
+
+            this._kernel?.emitJupyterLog(`Stopping Positron LSP server`);
+            await this._lsp.deactivate();
+        });
     }
 
     async restart(workingDirectory?: string): Promise<void> {
         if (this._kernel) {
-            // Stop the LSP client before restarting the kernel. Don't stop it
-            // until fully started to avoid an inconsistent state where the
-            // deactivation request comes in between the creation of the LSP
-            // comm and the LSP client.
-            //
-            // A cleaner way to set this up might be to put `this._lsp` in
-            // charge of creating the LSP comm, then `deactivate()` could
-            // keep track of this state itself.
-            await Promise.race([
-                this._lspStarting,
-                whenTimeout(400, () => {
-                    this._kernel!.emitJupyterLog('LSP startup timed out during interpreter restart');
-                }),
-            ]);
-            await this._lsp?.deactivate(true);
-
+            await this.deactivateLsp();
```

Comment:
@seeM did we decide we do want to keep `lspStarting` for here?

Reply:
Yes! I've reverted these changes